### ISSUE TITLE
Is a Linux remote builder required to initialize this?

### ISSUE THREAD

**Issue by spikespaz** [cpick/nix-rosetta-builder#7]
I am very excited to try this out. Unfortunately, it looks like the initial setup requires derivations for Linux to be built.

```
error: a 'aarch64-linux' with features {} is required to build '/nix/store/99km2c21rvw6z1vhkqwlrz8f24ldyvrr-X-Restart-Triggers-dhcpcd.drv', but I am a 'aarch64-darwin' with features {apple-virt, benchmark, big-parallel, nixos-test}
```

Must I configure another remote builder first?

Perhaps I can use [`nix.linux-builder`](https://daiderd.com/nix-darwin/manual/index.html#opt-nix.linux-builder.enable) to bootstrap this one. I will try it.

**Comment by spikespaz** [comment]
This does indeed work:
```nix
# The default builder is needed to bootstrap `nix-rosetta-builder`.
# Uncomment the line below, comment the import following, and switch.
# nix.linux-builder.enable = true;
# Then, uncomment this once you have switched with the `nix.linux-builder` enabled,
# comment out the above, and switch again.
imports = [
  inputs.nix-rosetta-builder.darwinModules.default
];
```

**Comment by cpick** [comment]
>  Must I configure another remote builder first?

Yes, unfortunately.

> Perhaps I can use [`nix.linux-builder`](https://daiderd.com/nix-darwin/manual/index.html#opt-nix.linux-builder.enable) to bootstrap this one. I will try it.

That's exactly what I'd recommend (and what you've done).  I'll add that to the README.md.

Bigger picture I'd love to remove this requirement.  The existing `nix.linux-builder` sidesteps this bootstrapping problem by being in the upstream NixOS hydra caches.
In fact, the nix-darwin [docs say](https://daiderd.com/nix-darwin/manual/index.html#opt-nix.linux-builder.config):
> This option specifies extra NixOS configuration for the builder. You should first use the Linux builder without changing the builder configuration otherwise you may not be able to build the Linux builder.

I don't actually think that this *should* be necessary.  I *think* the underlying reason for this requirement is that [internally building the image uses](https://discourse.nixos.org/t/passthru-config-to-help-debug-nixos-generators-nixosgenerate/57119) `pkgs.vmTools.runInLinuxVM`.  Naively, I don't see why this would be a firm requirement; all the contained binaries are already built and cached, it should be a "simple matter" of wrangling them into an image which could be done on aarch64-darwin, but I haven't confirmed whether this is true or how much work it'd really be.

A potentially easier option would be to also get this `rosetta-builder` upstreamed into Nixpkgs where it would also benefit from being pre-built and cached (in its default configuration, anyway).

**Comment by aspauldingcode** [comment]
I'm having initial issues when trying to rebuild with nix.linux-builder.enable = true; for the first time on a fresh install of nix-darwin. Looking forward to a fix which doesn't require this bootstrap method.

**Comment by aspauldingcode** [comment]
> I'm having initial issues when trying to rebuild with nix.linux-builder.enable = true; for the first time on a fresh install of nix-darwin. Looking forward to a fix which doesn't require this bootstrap method.

Found the issue potentially.
I didn't have a cache substituter. 
Others on nix-darwin have this issue:
https://github.com/LnL7/nix-darwin/issues/1081#issuecomment-2394824794

```
nix.linux-builder.enable = true;
nix.settings.substituters = [ "https://cache.nixos.org/" ];
nix.settings.trusted-public-keys = [ "hydra.nixos.org-1:CNHJZBh9K4tP3EKF6FkkgeVYsS3ohTl+oS0Qa8bezVs=" ];
nix.settings.always-allow-substitutes = true;
```


Then I can run a test example:

```
nix build \
  --impure \
  --expr '(with import <nixpkgs> { system = "aarch64-linux"; }; runCommand "foo" {} "uname -a > $out")'
  
~ cat result
Linux localhost 6.6.66 #1-NixOS SMP Sat Dec 14 19:00:23 UTC 2024 aarch64 GNU/Linux
```

**Comment by cpick** [comment]
> I'm having initial issues when trying to rebuild with nix.linux-builder.enable = true; for the first time on a fresh install of nix-darwin. Looking forward to a fix which doesn't require this bootstrap method.

> Found the issue potentially. I didn't have a cache substituter. Others on nix-darwin have this issue: [LnL7/nix-darwin#1081 (comment)](https://github.com/LnL7/nix-darwin/issues/1081#issuecomment-2394824794)

Yes, that would definitely do it.  And to be clear initial installation/bootstrapping will always require at either an existing Linux builder of some kind or a substituter that has `nix-rosetta-builder`s Linux requirements pre-cached.  Even if we find a way to make the final image natively on Darwin that image is filled with Linux binaries that still need to be compiled or substituted from a cache.

**Comment by cpick** [comment]
Some related discussion on Matrix:
https://matrix.to/#/!lheuhImcToQZYTQTuI:nixos.org/$Rsm3GqHH6K7ac7rwEW7Jb_aU3jcA_W4NWz5GcXKT1zI?via=nixos.org&via=matrix.org&via=nixos.dev

**Comment by dwt** [comment]
Is there a way to automatiacally fall back to the o.g. linux builder if the rosetta-builder is not yet available? My current working model of how nix works, is that it is functional, i.e. time _should_ not matter, so there should not be a way to check that the previous generation of the system already had something that is able to build linux binaries.

But... is that actually true?

**Comment by cpick** [comment]
> Is there a way to automatiacally fall back to the o.g. linux builder if the rosetta-builder is not yet available?

I haven't tested it, but you should be able to run both at the same time (they listen on different ports, etc).

**Comment by dwt** [comment]
Both at the same time is not what I tried to get. What I would like is that in the settings the first time you turn on the rosetta builder, it checks wether it was on before, and if its s not and therefore cannot build itself, it enables the linux builder one time to build itself, and then disables it again.

This is a very procedural thing that my working model of nix says should be hard to impossible. But then again, most functional languages have an out somewhere to achieve what is otherwise hard.

**Comment by cpick** [comment]
> What I would like is that in the settings the first time you turn on the rosetta builder, it checks wether it was on before, and if its s not and therefore cannot build itself, it enables the linux builder one time to build itself, and then disables it again.

I see, you're looking to automate [those initial installation steps](https://github.com/cpick/nix-rosetta-builder/blob/30dedb1c3c4be561e811cbf93791a21c7b082766/README.md?plain=1#L47-L52) within this module.  That's a good question that I haven't thought too much about.  I'd certainly be interested in ideas to make the new user's path easier.

**Comment by cpick** [comment]
> Could that perhaps enable the mechanism where the initial image is downloaded? Probably not, but I find the thought of reusing all those documented options quite intriguing. I do not understand however if this would prevent rosetta-builder from adding it's own like `linger-time`.

I think I follow your line of thought, but I'm not sure it gets us around needing an initial Linux builder of some type.

When someone enables the upstream [`nix.linux-builder.enable`](https://daiderd.com/nix-darwin/manual/index.html#opt-nix.linux-builder.enable) option (and leaves the rest of the `nix.linux-buider.*` options at their default) then they (usually) don't need any other, preexisting Linux builder for nix-darwin to install it.  This is because, I believe, it's built on the upstream [`darwin.linux-builder`](https://nixos.org/manual/nixpkgs/stable/#sec-darwin-builder) and Hydra builds and caches that for aarch64-darwin (including all the Linux binaries and images that it contains).  This means that as long as you don't change the builder's options in nix-darwin (as is noted in the docs for [`nix.linux-builder.config`](https://daiderd.com/nix-darwin/manual/index.html#opt-nix.linux-builder.config)) it should work fine.

So forming `nix-rosetta-builder` into the shape expected by `nix.linux-builder.package` may allow it to be installed via nix-darwin (it would still need the type of configuration that module.nix provides, though), but won't solve our bootstrapping problem (something would still need to build `nix-rosetta-builder`s image on Linux).

It *is* possible that we ultimately get `nix-rosetta-builder` merged into nixpkgs (and thus built by Hydra) which could allow for the same initial experience as the existing `linux-builder` (as long as no other settings were changed).  I want to get a little more stability and experience into this package before attempting that, but it seems reasonable.  I would *love* to find a way to build the VM's image without needing a Linux builder and that may be possible, but is a bigger engineering effort.

**Comment by dwt** [comment]
I just found out how [the upstream linux builder achieved the bootstrap before it became part of nixpkgs.](https://github.com/Gabriella439/macos-builder/blob/main/README.md)

Essentially what it boiled down to, is [this part of their flake](https://github.com/Gabriella439/macos-builder/blob/f21b8046f0060cdc3d7cfb90b39c397228d8353e/flake.nix#L188) where they add a cachix cache for their flake outputs, which contains the initial image:

```nix
  nixConfig = {
    extra-substituters = [ "https://macos-builder.cachix.org" ];

    extra-trusted-public-keys = [
      "macos-builder.cachix.org-1:HPWcq59/iyqQz6HEtlO/kjD/a7ril0+/XJc+SZ2LgpI="
    ];
  };
```

Then that was no longer necessary once hydra built the image instead.

**Comment by cpick** [comment]
That's an interesting solution (and thank you for digging it up!).
I'm a little nervous about doing that by default (I don't like adding trusted substituters to my machines), but I think it's an idea that can be added as an option (possibly by default after it becomes comfortable).

I still hope that ultimately there will be a way to create a VM image without needing a Linux machine, but that's likely a bigger project.

**Comment by dwt** [comment]
From my experience with nixpkgs, it is probably not too hard to get the initial image upstream, if you feel it is stable enough. That would also solve the problem, and would just slow down changes to the initial image.

That being said, as long as the initial image is able to build the next one, that could be ok?
